File: sdk-js/test/content.test.ts
import { beforeAll, describe, expect, it } from "vitest";
import { GlyphuxClient } from "../src/client.js";
import { GlyphuxApiError } from "../src/errors.js";
import { testEnv } from "./testenv.js";

// The daemon's composition (seeded by sdk-js/testdata/seedcomposition, run
// from test/global-setup.ts) declares one content type, "post", with fields
// title (string, required), body (string), featured (boolean), and
// hero_image (media).
async function adminClient(): Promise<GlyphuxClient> {
  const env = testEnv();
  const client = new GlyphuxClient({ baseUrl: env.baseUrl });
  await client.auth.login(env.adminEmail, env.adminPassword);
  return client;
}

describe("content", () => {
  it("create() persists a new draft item at version 1", async () => {
    const client = await adminClient();

    const item = await client.content.create("post", { title: "Hello world" });

    expect(item.type).toBe("post");
    expect(item.data).toEqual({ title: "Hello world" });
    expect(item.status).toBe("draft");
    expect(item.version).toBe(1);
    expect(typeof item.id).toBe("string");
    expect(item.id.length).toBeGreaterThan(0);
  });

  it("create() rejects an unknown content type as a typed 404 error", async () => {
    const client = await adminClient();

    let caught: unknown;
    try {
      await client.content.create("no-such-type", { title: "x" });
    } catch (err) {
      caught = err;
    }

    expect(caught).toBeInstanceOf(GlyphuxApiError);
    expect((caught as InstanceType<typeof GlyphuxApiError>).status).toBe(404);
  });

  it("create() reports validation failures as a 422 with field issues", async () => {
    const client = await adminClient();

    let caught: unknown;
    try {
      // "post" requires "title" (see composition seeded above).
      await client.content.create("post", { body: "no title here" });
    } catch (err) {
      caught = err;
    }

    const err = caught as InstanceType<typeof GlyphuxApiError>;
    expect(err).toBeInstanceOf(GlyphuxApiError);
    expect(err.status).toBe(422);
    expect(err.issues).toBeDefined();
    expect(err.issues!.length).toBeGreaterThan(0);
    expect(err.issues![0]).toContain("title");
  });

  it("get() returns a previously created item by id", async () => {
    const client = await adminClient();
    const created = await client.content.create("post", { title: "Fetch me" });

    const fetched = await client.content.get("post", created.id);

    expect(fetched.id).toBe(created.id);
    expect(fetched.data).toEqual({ title: "Fetch me" });
  });

  it("get() reports a missing item as a typed 404 error", async () => {
    const client = await adminClient();

    let caught: unknown;
    try {
      await client.content.get("post", "no-such-id");
    } catch (err) {
      caught = err;
    }

    expect((caught as InstanceType<typeof GlyphuxApiError>).status).toBe(404);
  });

  it("list() includes a created item for its type", async () => {
    const client = await adminClient();
    const created = await client.content.create("post", { title: "Listed item" });

    const items = await client.content.list("post");

    expect(items.some((i) => i.id === created.id)).toBe(true);
  });

  it("update() overwrites the item's data and bumps its version", async () => {
    const client = await adminClient();
    const created = await client.content.create("post", { title: "Original" });

    const updated = await client.content.update("post", created.id, { title: "Changed" });

    expect(updated.data).toEqual({ title: "Changed" });
    expect(updated.version).toBe(2);
  });

  it("delete() removes the item so a later get() 404s", async () => {
    const client = await adminClient();
    const created = await client.content.create("post", { title: "Doomed" });

    await client.content.delete("post", created.id);

    let caught: unknown;
    try {
      await client.content.get("post", created.id);
    } catch (err) {
      caught = err;
    }
    expect((caught as InstanceType<typeof GlyphuxApiError>).status).toBe(404);
  });

  it("publish() then unpublish() toggle an item's status", async () => {
    const client = await adminClient();
    const created = await client.content.create("post", { title: "Publish me" });

    const published = await client.content.publish("post", created.id);
    expect(published.status).toBe("published");

    const unpublished = await client.content.unpublish("post", created.id);
    expect(unpublished.status).toBe("draft");
  });

  it("listVersions() and rollback() expose an item's version history", async () => {
    const client = await adminClient();
    const created = await client.content.create("post", { title: "v1" });
    await client.content.update("post", created.id, { title: "v2" });

    const versions = await client.content.listVersions("post", created.id);
    expect(versions.map((v) => v.version).sort()).toEqual([1, 2]);
    expect(versions.find((v) => v.version === 1)?.data).toEqual({ title: "v1" });

    const rolledBack = await client.content.rollback("post", created.id, 1);
    expect(rolledBack.data).toEqual({ title: "v1" });
    expect(rolledBack.version).toBe(3);
  });
});

File: pkg/client/content_test.go
package client_test

import (
	"context"
	"encoding/json"
	"testing"

	"github.com/glyphux/glyphux/pkg/client"
)

func loggedInClient(t *testing.T) *client.Client {
	t.Helper()
	ts, identities, _ := newTestServer(t)
	ctx := context.Background()
	if err := identities.CreateAdmin(ctx, "admin@example.com", "correct horse battery"); err != nil {
		t.Fatal(err)
	}
	c := client.New(ts.URL)
	if _, err := c.Auth.Login(ctx, "admin@example.com", "correct horse battery"); err != nil {
		t.Fatalf("Login: %v", err)
	}
	return c
}

func TestContentCreateAndGet(t *testing.T) {
	c := loggedInClient(t)
	ctx := context.Background()

	created, err := c.Content.Create(ctx, "article", map[string]any{"title": "Hello", "body": "World"})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if created.ID == "" {
		t.Fatal("Create: item has no id")
	}
	if created.Data["title"] != "Hello" {
		t.Errorf("Create: data[title] = %v, want Hello", created.Data["title"])
	}
	if created.Status != "draft" {
		t.Errorf("Create: status = %q, want draft", created.Status)
	}

	got, err := c.Content.Get(ctx, "article", created.ID)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if got.Data["title"] != "Hello" {
		t.Errorf("Get: data[title] = %v, want Hello", got.Data["title"])
	}
}

func TestContentListUpdateDelete(t *testing.T) {
	c := loggedInClient(t)
	ctx := context.Background()

	created, err := c.Content.Create(ctx, "article", map[string]any{"title": "Hello", "body": "World"})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}

	items, err := c.Content.List(ctx, "article")
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if len(items) != 1 || items[0].ID != created.ID {
		t.Fatalf("List = %+v, want one item with id %q", items, created.ID)
	}

	updated, err := c.Content.Update(ctx, "article", created.ID, map[string]any{"title": "Updated", "body": "World"})
	if err != nil {
		t.Fatalf("Update: %v", err)
	}
	if updated.Data["title"] != "Updated" {
		t.Errorf("Update: data[title] = %v, want Updated", updated.Data["title"])
	}

	if err := c.Content.Delete(ctx, "article", created.ID); err != nil {
		t.Fatalf("Delete: %v", err)
	}
	if _, err := c.Content.Get(ctx, "article", created.ID); err == nil {
		t.Fatal("Get after Delete: want error, got nil")
	}
}

func TestContentPublishUnpublishVersionsRollback(t *testing.T) {
	c := loggedInClient(t)
	ctx := context.Background()

	created, err := c.Content.Create(ctx, "article", map[string]any{"title": "v1", "body": "World"})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}

	published, err := c.Content.Publish(ctx, "article", created.ID)
	if err != nil {
		t.Fatalf("Publish: %v", err)
	}
	if published.Status != "published" {
		t.Errorf("Publish: status = %q, want published", published.Status)
	}

	if _, err := c.Content.Update(ctx, "article", created.ID, map[string]any{"title": "v2", "body": "World"}); err != nil {
		t.Fatalf("Update: %v", err)
	}

	versions, err := c.Content.ListVersions(ctx, "article", created.ID)
	if err != nil {
		t.Fatalf("ListVersions: %v", err)
	}
	if len(versions) < 2 {
		t.Fatalf("ListVersions = %+v, want at least 2 versions", versions)
	}

	rolled, err := c.Content.Rollback(ctx, "article", created.ID, 1)
	if err != nil {
		t.Fatalf("Rollback: %v", err)
	}
	if rolled.Data["title"] != "v1" {
		t.Errorf("Rollback: data[title] = %v, want v1", rolled.Data["title"])
	}

	unpublished, err := c.Content.Unpublish(ctx, "article", created.ID)
	if err != nil {
		t.Fatalf("Unpublish: %v", err)
	}
	if unpublished.Status != "draft" {
		t.Errorf("Unpublish: status = %q, want draft", unpublished.Status)
	}
}

func TestContentCreateValidationFailureExposesIssues(t *testing.T) {
	c := loggedInClient(t)
	ctx := context.Background()

	// "title" is required on the article type (see newTestServer); omitting
	// it must fail validation with the field issue surfaced on the error.
	_, err := c.Content.Create(ctx, "article", map[string]any{"body": "World"})
	var apiErr *client.APIError
	if !client.AsAPIError(err, &apiErr) {
		t.Fatalf("Create with missing required field error = %v (%T), want *client.APIError", err, err)
	}
	if apiErr.StatusCode != 422 {
		t.Fatalf("Create validation error status = %d, want 422", apiErr.StatusCode)
	}
	var issues []string
	if err := json.Unmarshal(apiErr.Issues, &issues); err != nil {
		t.Fatalf("decode APIError.Issues: %v (raw %s)", err, apiErr.Issues)
	}
	if len(issues) == 0 {
		t.Fatal("APIError.Issues is empty, want at least one validation issue")
	}
}

func TestContentGetUnknownIDReturnsAPIError(t *testing.T) {
	c := loggedInClient(t)
	ctx := context.Background()

	_, err := c.Content.Get(ctx, "article", "does-not-exist")
	var apiErr *client.APIError
	if !client.AsAPIError(err, &apiErr) {
		t.Fatalf("Get unknown id error = %v (%T), want *client.APIError", err, err)
	}
	if apiErr.StatusCode != 404 {
		t.Errorf("Get unknown id status = %d, want 404", apiErr.StatusCode)
	}
}
